Add PostgreSQL auto-user deletion

gabrielcorado opened the pull request:
Related to #31199 and #29731

Adds a create database user mode (similar to what is done for host and desktop users), where:
- `off`: Disables user creation. (same as `create_database_user: false`).
- `keep`: Creates the user but deactivates it when the session ends (keeps the user but removes its roles assignments). This is the current default behavior.
- `best_effort_drop`: Tries to drop the database user. If it fails, it returns to `keep` mode, deactivating the user.

For backward compatibility, the `create_database_user: true` role option is kept, and set the mode value to `keep` if it is set to `true`.

gabrielcorado committed:
feat: add auto-user deletion postgres

greedy52 commented on lib/srv/db/common/autousers.go:
```diff
@@ -51,7 +51,7 @@ type UserProvisioner struct {
 // function to make sure no 2 processes run user activation simultaneously.
 func (a *UserProvisioner) Activate(ctx context.Context, sessionCtx *Session) (func(), error) {
-	if !sessionCtx.AutoCreateUser {
+	if !services.IsCreateDatabaseUserEnabled(sessionCtx.AutoCreateUserMode) {
```
I would throw errors if the Backend doesn't support the new mode yet (somewhere here). IMO, it's better to alert the user the mode configuration is wrong than silently falling back to another mode. But certainly up for discussion.

greedy52 commented on lib/srv/db/postgres/engine.go:
```diff
@@ -574,6 +574,22 @@ func (e *Engine) handleCancelRequest(ctx context.Context, sessionCtx *common.Ses
 
+// teardownUser chooses and call the auto provisioner method used to drop
+// auto-users.
+func (e *Engine) teardownUser(ctx context.Context, sessionCtx *common.Session) error {
```
should this be moved to `common.UserProvisioner`? Seems like common logic to me.

greedy52 commented on lib/srv/db/postgres/delete-user.sql:
```diff
@@ -0,0 +1,20 @@
+        begin
+            execute format('drop user %I', username);
+        exception
+            when SQLSTATE '2BP01' then
+                -- Drop user/role will fail if user has dependent objects.
+                -- In this scenario, fallback into disabling the user.
+                call teleport_deactivate_user(username);
```
curious if we are able to tell if user is dropped successfully or the user is deactivated based on our logs?

gabrielcorado replied:
Initially, I've added a specific return for delete and deactivate. I've removed it since this procedure is simple and only has a fallback when a particular error happens. I can bring it back if you feel this would have a good debugging value.

greedy52 replied:
> I can bring it back if you feel this would have a good debugging value.

That would be great. I think it will be valuable to tell if `preferDrop` actually drops or not.

gabrielcorado commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -2453,6 +2453,19 @@ enum CreateHostUserMode {
+  DB_USER_MODE_OFF = 1;
+  // DB_USER_MODE_KEEP allows user creation and disable users at session end.
+  DB_USER_MODE_KEEP = 2;
+  // DB_USER_MODE_PREFER_DROP allows user creation and prefers dropping users
+  // if they don't own any resources. Otherwise, it disables them.
+  DB_USER_MODE_PREFER_DROP = 3;
```
My thought is to make these modes work the same way on each engine, making it easier for users to understand what is going to be the effect of each mode. This could lead to modes not making much sense in some engines, so you might be right about it being specific.

Note: This is not necessarily a limitation on the PostgreSQL side. It is just more trivial than other engines. We could use the `DROP OWNED` to drop the user objects or call the `REASSIGN OWNED` to change their owner.

> MySQL and MongoDB don't have ownership so users can be dropped without any issue.

Could a user create a table that only they can access? Or do their roles guarantee other users access to it?

I wonder if there could be any advantage to using the "prefer drop" for them.

espadolini replied:
best_effort_drop?

greedy52 replied:
> Could a user create a table that only they can access?

No. The table is owned by the database, not the user who produces it (communism?).

> My thought is to make these modes work the same way on each engine, making it easier for users to understand what is going to be the effect of each mode. This could lead to modes not making much sense in some engines, so you might be right about it being specific.

Ok. So will MySQL support a separate `drop` mode besides `preferDrop`?

gabrielcorado replied:
> No. The table is owned by the database, not the user who produces it (communism?).

I see. However, there is a case where there can be a problem: when the user creates procedures. The drop command only returns a warning, but when anyone calls the procedure, it returns an error. (This also applies to views).

The point is: to make a safe version of the drop command, and we could discuss having another version that could drop users anyway (or execute the necessary commands to ensure it will succeed).

> Ok. So will MySQL support a separate drop mode besides preferDrop?

If we think a `drop` makes sense, yes. Then all the engines would need to implement it also.

gabrielcorado replied:
> best_effort_drop?

That sounds like a better option. And since we already have this "format" on other parts, it would be better to keep it.

gabrielcorado commented on lib/srv/db/common/autousers.go:
```diff
@@ -131,6 +131,35 @@ func (a *UserProvisioner) Deactivate(ctx context.Context, sessionCtx *Session) e
 
+// Delete deletes a database user.
+func (a *UserProvisioner) Delete(ctx context.Context, sessionCtx *Session) error {
+	if !services.IsCreateDatabaseUserEnabled(sessionCtx.AutoCreateUserMode) {
+		return nil
+	}
```
I'll change it to make it more consistent, but the [engines already match the mode to call the correct teardown function](https://github.com/gravitational/teleport/blob/5fda5c9961608f9e360e658f77e7066ee8e9cde9/lib/srv/db/postgres/engine.go#L579-L591).

gabrielcorado commented on lib/srv/db/autousers_test.go:
```diff
@@ -30,49 +30,81 @@ import (
+			databaseRoles:       []string{"reader", "writer"},
+			expectConnectionErr: false,
+		},
+		"disabled": {
+			mode:                types.CreateDatabaseUserMode_DB_USER_MODE_OFF,
+			databaseRoles:       []string{"reader", "writer"},
+			expectConnectionErr: true,
+		},
```
I'll add a comment to it. But this should fail because the username is absent on the database and not created by the procedure, so it returns an access denied error.

gabrielcorado commented on lib/srv/db/common/test.go:
```diff
@@ -69,6 +74,11 @@ func (cfg *TestServerConfig) CheckAndSetDefaults() error {
 	}
 
+	if cfg.Users == nil {
+		cfg.Users = make([]string, 0)
```
Good catch. This was a left over.

gabrielcorado commented on lib/services/access_checker.go:
```diff
@@ -514,40 +514,41 @@ func (a *accessChecker) Traits() wrappers.Traits {
 	for _, role := range a.RoleSet {
-		if role.GetCreateDatabaseUserOption() {
+		mode := role.GetCreateDatabaseUserMode()
+		if mode != types.CreateDatabaseUserMode_DB_USER_MODE_UNSPECIFIED && mode != types.CreateDatabaseUserMode_DB_USER_MODE_OFF {
```
I've tried to keep the same behavior from using the `create_database_user` option. If there is a role (allow) with it set to true, it will always create the user.

We could update it to match this "pattern", but it would lead to breaking changes.

gabrielcorado committed:
refactor: change to IsEnabled func to check auto-user

gabrielcorado committed:
test: fix linting and test

gabrielcorado committed:
refactor(db): code review suggestions

greedy52 reviewed (approved):
LGTM, assuming a debug log will be added somewhere to show if the user is actually deleted or a fallback is done.

gabrielcorado commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -2614,6 +2627,10 @@ message RoleOptions {
+
+  // CreateDatabaseUserMode allows users to be automatically created on a
+  // database when not set to off.
+  CreateDatabaseUserMode CreateDatabaseUserMode = 29 [(gogoproto.jsontag) = "create_database_user_mode,omitempty"];
```
I see. I'll rename it to the short version to keep the same pattern as the previous configuration (`create_db_user`).

gabrielcorado commented on lib/services/access_checker.go:
```diff
@@ -514,40 +514,41 @@ func (a *accessChecker) Traits() wrappers.Traits {
 	for _, role := range a.RoleSet {
-		if role.GetCreateDatabaseUserOption() {
+		mode := role.GetCreateDatabaseUserMode()
+		if mode != types.CreateDatabaseUserMode_DB_USER_MODE_UNSPECIFIED && mode != types.CreateDatabaseUserMode_DB_USER_MODE_OFF {
```
Yes. I'll update it.

gabrielcorado committed:
refactor: rename option to best effort drop

gabrielcorado committed:
refactor(api): rename createa database user mode property

gabrielcorado committed:
refactor(services): review suggestions

gabrielcorado committed:
feat(postgres): add log for user deletion result

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/pg-user-auto-deletion

gabrielcorado commented:
@greedy52 @smallinsky FYI, I've made the following updates:
- Changed `preferDrop` mode name to `best_effort_drop` (keeping the same pattern used in some other places).
- Added logs into `DeleteUser` indicating if the user was dropped (debug) or deactivated (info).

gabrielcorado committed:
refactor(integrations): regenerate crd manifests

greedy52 commented on lib/srv/db/postgres/users.go:
```diff
@@ -163,6 +173,13 @@ const (
+	sqlStateUserDropped = "TP001"
+	// sqlStateUserDropped is the state returned by the delete procedure
+	// indicating was deactivated.
+	sqlStateUserDeactivated = "TP002"
```
Could we use the same state sas MySQL? Probably can move these state constants to common too.

gabrielcorado replied:
Done on commit [4cf70ef0e8f422e59f3831816d6af88439f2c2c8](https://github.com/gravitational/teleport/pull/32792/commits/4cf70ef0e8f422e59f3831816d6af88439f2c2c8).

greedy52 replied:
Thanks!!

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/pg-user-auto-deletion

gabrielcorado committed:
feat(examples): update operator role spec

gabrielcorado committed:
refactor(db): use common sql state codes
